### PR TITLE
More optimizer refactoring: move eigensolver to another file

### PR DESCRIPTION
**Pull request by markdewing**
Simple refactorings:
* Move the eigensolver code to its own file.
* Also do a clang format missed in the previous PR.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
desktop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- N/A. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- N/A. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/WFOpt/Eigensolver.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  int lwork(-1);
+  std::vector<Real> work(1);
+  Real tt(0);
+  int t(1);
+  LAPACK::ggev(&jl, &jr, &Nl, A.data(), &Nl, B.data(), &Nl, &alphar[0], &alphai[0], &beta[0], &tt, &t,
+               eigenvectors.data(), &Nl, &work[0], &lwork, &info);
+  lwork = int(work[0]);
+  work.resize(lwork);
+
+  LAPACK::ggev(&jl, &jr, &Nl, A.data(), &Nl, B.data(), &Nl, &alphar[0], &alphai[0], &beta[0], &tt, &t,
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/Eigensolver.cpp:66]: Could you change all the `&vec[0]` to `vec.data()`?

File: src/QMCDrivers/WFOpt/Eigensolver.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+               eigenvectors.data(), &Nl, &work[0], &lwork, &info);
+  if (info != 0)
+    throw std::runtime_error("Invalid Matrix Diagonalization Function, ggev info = " + std::to_string(info));
+
+  for (int i = 0; i < Nl; i++)
+  {
+    eigenvals[i] = alphar[i] / beta[i];
+  }
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/Eigensolver.cpp:74]: remove `{}` around one line.

File: src/QMCDrivers/WFOpt/Eigensolver.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  work.resize(lwork);
+
+  LAPACK::geev(&jl, &jr, &Nl, prdMat.data(), &Nl, &alphar[0], &alphai[0], eigenD.data(), &Nl, eigenvectors.data(), &Nl,
+               &work[0], &lwork, &info);
+  if (info != 0)
+    throw std::runtime_error("Invalid Matrix Diagonalization Function, geev info = " + std::to_string(info));
+
+  for (int i = 0; i < Nl; i++)
+  {
+    eigenvals[i] = alphar[i];
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/Eigensolver.cpp:123]: Should we use eigenvals directly in the geev call?

File: src/QMCDrivers/WFOpt/Eigensolver.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+
+/** @file EigenSolver.h
+ * @brief EigenSolver related functions.
+ */
+#ifndef QMCPLUSPLUS_EIGENSOLVER_H
+#define QMCPLUSPLUS_EIGENSOLVER_H
+
+#include <Configuration.h>
+#include <NewTimer.h>
```
[the remark this answers is not in the document]
Comment (markdewing) [reply src/QMCDrivers/WFOpt/Eigensolver.h]: Removed. It's not used in the cpp file.

File: src/QMCDrivers/WFOpt/Eigensolver.cpp (hunk 1 above)
Comment (markdewing) [reply src/QMCDrivers/WFOpt/Eigensolver.cpp]: Fixed.

File: src/QMCDrivers/WFOpt/Eigensolver.cpp (hunk 2 above)
Comment (markdewing) [reply src/QMCDrivers/WFOpt/Eigensolver.cpp]: Fixed.

File: src/QMCDrivers/WFOpt/Eigensolver.cpp (hunk 3 above)
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/Eigensolver.cpp:120]: @markdewing

Comment (markdewing) [reply src/QMCDrivers/WFOpt/Eigensolver.cpp:120]: Sure, I'll make that change.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**